src/Core/Managed/Shared/Extensibility/MetricManager.cs

```diff
@@ -0,0 +1,285 @@
+                statistics.Count,
+                statistics.Sum,
+                statistics.Min,
+                statistics.Max,
+                statistics.StandardDeviation);
+
+            if (metric.Dimensions != null)
+            {
+                foreach (KeyValuePair<string, string> property in metric.Dimensions)
+                {
+                    telemetry.Properties.Add(property);
+                }
+            }
+
+            return telemetry;
+        }
+
+        private void Flush(bool disposing)
+        {
+            if ((!disposing) && this.disposed)
+            {
+                throw new ObjectDisposedException(nameof(MetricManager));
+            }
+
+            try
+            {
+                this.Snapshot();
+                this.telemetryClient.Flush();
+            }
+            catch (Exception ex)
+            {
+                CoreEventSource.Log.FailedToFlushMetricAggregators(ex.ToString());
+            }
+        }
+
+        /// <summary>
+        /// Takes a snapshot of aggregators collected by this instance of the manager
+        /// and schedules the next snapshot.
+        /// </summary>
+        private Task SnapshotAndReschedule()
+        {
+            return Task.Factory.StartNew(
+                () =>
+                {
+                    try
+                    {
+                        this.Snapshot();
+                    }
+                    catch (Exception ex)
+                    {
+                        CoreEventSource.Log.FailedToSnapshotMetricAggregators(ex.ToString());
+                    }
+                    finally
+                    {
+                        this.snapshotTimer.Delay = GetWaitTime();
+                        this.snapshotTimer.Start(this.SnapshotAndReschedule);
+                    }
+                });
+        }
+
+        /// <summary>
+        /// Takes snapshot of all active metric aggregators and turns results into metric telemetry.
+        /// </summary>
+        private void Snapshot()
+        {
+            ConcurrentDictionary<Metric, SimpleMetricStatisticsAggregator> aggregatorSnapshot =
+                Interlocked.Exchange(ref this.metricDictionary, new ConcurrentDictionary<Metric, SimpleMetricStatisticsAggregator>());
+
+            // calculate aggregation interval duration interval
+            TimeSpan aggregationIntervalDuation = DateTimeOffset.UtcNow - this.lastSnapshotStartDateTime;
+            this.lastSnapshotStartDateTime = DateTimeOffset.UtcNow;
+
+            // prevent zero duration for interval
+            if (aggregationIntervalDuation.TotalMilliseconds < 1)
+            {
+                aggregationIntervalDuation = TimeSpan.FromMilliseconds(1);
+            }
+
+            // adjust interval duration to exactly snapshot frequency if it is close (within 1%)
+            double difference = Math.Abs(aggregationIntervalDuation.TotalMilliseconds - snapshotFrequency.TotalMilliseconds);
+
```

Comment:
So it the period is 1 minute, 1% is 0.6 secs. For 5 mins aggregations, it is 3 secs. I am not sure that this is what you want. Maybe 1% for under period < 10 secs; 0.5 sec for 10 sec < period <= 30 sec; 1 sec for 30 sec < period  <= 3 min; and 3 secs otherwise. This is also random, bu i think it comes closer to what you want. Whaever you choose, please explain it in a comment.